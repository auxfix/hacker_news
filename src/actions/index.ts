export * from './alerts';
export * from 'feature/news/actions';
