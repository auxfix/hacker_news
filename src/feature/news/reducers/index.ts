export { newsState, default as news } from './news';
