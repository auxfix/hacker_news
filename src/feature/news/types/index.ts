export interface hackerstory {
  authorid: string;
  id: number;
  img?: string;
  karma?: number;
  score: number;
  time: number;
  title: string;
  url: string;
}

export interface user {
  about: string;
  created: number;
  id: string;
  karma: number;
}
