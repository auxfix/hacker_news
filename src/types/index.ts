export * from './common';
export * from './state';
export * from 'feature/news/types';
