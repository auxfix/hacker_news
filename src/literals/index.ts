import { keyMirror } from '@gilbarbara/helpers';
import { Status } from 'types';

export const ActionTypes = keyMirror({
  HIDE_ALERT: undefined,
  SHOW_ALERT: undefined,
  SET_APP_OPTIONS: undefined,

  // news
  GET_NEWS_REQUEST: undefined,
  GET_NEWS_SUCCESS: undefined,
  GET_NEWS_FAILURE: undefined,
  GET_NEWS_PIECE_SUCCESS: undefined,
});

export const STATUS: Status = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
