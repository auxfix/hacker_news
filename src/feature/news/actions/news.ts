import { createAction } from '@reduxjs/toolkit';
import { ActionTypes } from 'literals';
import { actionPayload } from 'modules/helpers';
import { HackerStory } from 'types';

export const getNews = createAction(ActionTypes.GET_NEWS_REQUEST);
export const getNewsSuccess = createAction(ActionTypes.GET_NEWS_SUCCESS, (payload: HackerStory[]) =>
  actionPayload(payload),
);
export const getNewsFailure = createAction(ActionTypes.GET_NEWS_FAILURE);
