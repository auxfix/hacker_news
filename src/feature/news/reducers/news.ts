import { createReducer } from '@reduxjs/toolkit';
import { getNews, getNewsFailure, getNewsSuccess } from 'feature/news/actions';
import { NewsState } from 'types';

export const newsState: NewsState = {
  news: [],
  isLoading: false,
};

export default {
  news: createReducer<NewsState>(newsState, builder => {
    builder
      .addCase(getNews, draft => {
        draft.isLoading = true;
      })
      .addCase(getNewsSuccess, (draft, { payload }) => {
        draft.news = payload;
        draft.isLoading = false;
      })
      .addCase(getNewsFailure, draft => {
        draft.isLoading = false;
      });
  }),
};
