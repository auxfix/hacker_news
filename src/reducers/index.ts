import alerts, { alertsState } from './alerts';
import app, { appState } from './app';
import user, { userState } from './user';
import news, { newsState } from 'feature/news/reducers';

export const initialState = {
  alerts: alertsState,
  app: appState,
  user: userState,
  news: newsState,
};

export default {
  ...alerts,
  ...app,
  ...user,
  ...news
};
