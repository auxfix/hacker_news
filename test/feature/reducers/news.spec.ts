import { getNews, getNewsFailure, getNewsSuccess } from 'feature/news/actions/news';
import { news as reducer } from 'feature/news/reducers';
import { ActionTypes } from 'literals';
import { emptyAction } from 'test-utils';
import { HackerStory } from 'types';

const testNewsItem: HackerStory = {
  authorId: 'hacker.one',
  id: 242342343,
  score: 74,
  time: 1669726441,
  title: 'amazing article',
  url: 'https://en.wikipedia.org/wiki/Hacker',
  img: 'https://images/nice_image.jpg',
  karma: 34,
};

describe('reducers/feture/news', () => {
  let news = reducer.news(undefined, emptyAction);

  it('should return the initial state', () => {
    expect(reducer.news(news, emptyAction)).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.GET_NEWS_REQUEST}`, () => {
    news = reducer.news(news, getNews());
    expect(news).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.GET_NEWS_SUCCESS}`, () => {
    news = reducer.news(news, getNewsSuccess([testNewsItem]));
    expect(news).toMatchSnapshot();
  });

  it(`should handle ${ActionTypes.GET_NEWS_FAILURE}`, () => {
    news = reducer.news(news, getNewsFailure());
    expect(news).toMatchSnapshot();
  });
});
