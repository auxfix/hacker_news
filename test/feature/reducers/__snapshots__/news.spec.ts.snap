// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/feture/news should handle GET_NEWS_FAILURE 1`] = `
{
  "isLoading": false,
  "news": [
    {
      "authorId": "hacker.one",
      "id": 242342343,
      "img": "https://images/nice_image.jpg",
      "karma": 34,
      "score": 74,
      "time": 1669726441,
      "title": "amazing article",
      "url": "https://en.wikipedia.org/wiki/Hacker",
    },
  ],
}
`;

exports[`reducers/feture/news should handle GET_NEWS_REQUEST 1`] = `
{
  "isLoading": true,
  "news": [],
}
`;

exports[`reducers/feture/news should handle GET_NEWS_SUCCESS 1`] = `
{
  "isLoading": false,
  "news": [
    {
      "authorId": "hacker.one",
      "id": 242342343,
      "img": "https://images/nice_image.jpg",
      "karma": 34,
      "score": 74,
      "time": 1669726441,
      "title": "amazing article",
      "url": "https://en.wikipedia.org/wiki/Hacker",
    },
  ],
}
`;

exports[`reducers/feture/news should return the initial state 1`] = `
{
  "isLoading": false,
  "news": [],
}
`;
